EIPS/eip-152.md

```diff
@@ -0,0 +1,228 @@
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
+Besides being a useful cryptographic hash function and SHA3 finalist, BLAKE2b allows for efficient verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum. A single verification of an Equihash PoW verification requires 512 iterations of the hash function, making verification of Zcash block headers prohibitively expensive if a Solidity implementation of BLAKE2b is used.
+
+The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors.
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+We propose adding a precompiled contract at address `0x09` wrapping the [BLAkE2b `F` compression function](https://tools.ietf.org/html/rfc7693#section-3.2).
+
+The precompile requires 6 inputs tightly encoded, taking exactly 213 bytes, as explained below. The encoded inputs are corresponding to the ones specified in the [BLAKE2b RFC Section 3.2](https://tools.ietf.org/html/rfc7693#section-3.2):
+
+- `rounds` - the number of rounds
+- `h` - the state vector
+- `m` - the message block vector
+- `t_0, t_1` - offset counters
+- `f` - the final block indicator flag
+
+```
+[4 bytes for rounds][64 bytes for h][128 bytes for m][8 bytes for t_0][8 bytes for t_1][1 byte for f]
+```
+
+The boolean `f` parameter is considered as `true` if at least 1 bit out of the 8 bits is set.
+
+The precompile should compute the `F` function as [specified in the RFC](https://tools.ietf.org/html/rfc7693#section-3.2) and return the updated state vector `h`.
+
+### Example Usage in Solidity
+
+The precompile can be wrapped easily in Solidity to provide a more development-friendly interface to `F`.
+
+```solidity
+function F(uint32 rounds, bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f) public view returns (bytes32[2] memory) {
+  bytes32[2] memory output;
+
+  bytes memory args = abi.encodePacked(rounds, h[0], h[1], m[0], m[1], m[2], m[3], t[0], t[1], f);
```

Comment:
There appears to be a byte order problem here. BLAKE2b is consistently little-endian. `abi.encodePacked` encodes each of its arguments in big-endian order. EVM's memory accesses are big-endian for types larger than a byte.

For the `bytes32` values the intent is clear; a byte array doesn't have endianness, and so the input should be read in the same order as any BLAKE2b implementation operating on byte arrays would read it. However, technically there's a specification ambiguity because in [RFC 7693](https://tools.ietf.org/html/rfc7693#section-3.2), the `h` and `m` inputs to `F` are specified as being arrays of unsigned 32-bit words. They obviously should be read as little-endian, as the full BLAKE2b hash requires (the ambiguity is caused by the fact that this conversion is specified as part of the full hash, not `F`).

For the `uint64` values, on the other hand, it's totally unclear what the intent is. Note that RFC 7693 specifies `t` as a single integer, and unpacks it into words within the definition of `F`. I would imagine that users of this API would try to just increment `t[0]` for each block, and then using `abi.encodePacked` here would be plain wrong (the byte order would be swapped).

As well as clarifying this in the text, I strongly suggest adding test vectors (which should include `t` values >= 2<sup>64</sup>).

Reply:
Thanks for the review @daira! The test vectors in [the implementation](https://github.com/keep-network/go-ethereum/pull/4) seem to be working, derived from the golang Blake2b implementation and the RFC... so I suspect this would've been a "works on accident" type bug we found on another client implementation.

We'll clarify the text here to remove the ambiguity and see about a couple test vectors for the EIP on top of the geth implementation's vectors.

Thanks again for catching this!

Reply:
I'm confused as to how that implementation passes test vectors from the RFC. First of all there is only one execution trace for BLAKE2b in the RFC (in Appendix A), and the test vectors in `blake2FTests` do not include it in any byte order. Second, there are byte-reversed inputs in the test vectors &mdash; for example, "07060504030201000f0e0d0c0b0a09081716151413121110..." looks very much like it is derived from "000102030405060708090a0b0c0d0e0f1011121314151617..." with the order of bytes reversed in each 64-bit word, exactly as I'd expect if the implementation were using the wrong byte order.

As far as I can see, all of the conversions except for `rounds` (i.e. from line 404 to 428) should be little-endian. https://github.com/keep-network/go-ethereum/pull/4/files#diff-1d2b8390d9dc5764156c9dbe6358a30bR404

Reply:
My mistake, I meant to refer to the underlying [Blake2b library fork](https://github.com/keep-network/blake2b) which AFAIK is [using canonical test vectors](https://github.com/keep-network/blake2b/blob/04b7789d5c6a323c92a5baf6d1ef8039cd6cb404/hash_test.go#L282) that cover the `F` implementation (via the Blake2b tests, not `F` specific).

Reply:
Many thanks for the review @daira, I was quietly hoping to see someone from Zcash here.  🙇 

> There appears to be a byte order problem here. BLAKE2b is consistently little-endian. `abi.encodePacked` encodes each of its arguments in big-endian order. EVM's memory accesses are big-endian for types larger than a byte.
> 
> For the `bytes32` values the intent is clear; a byte array doesn't have endianness, and so the input should be read in the same order as any BLAKE2b implementation operating on byte arrays would read it. However, technically there's a specification ambiguity because in [RFC 7693](https://tools.ietf.org/html/rfc7693#section-3.2), the `h` and `m` inputs to `F` are specified as being arrays of unsigned 32-bit words. They obviously should be read as little-endian, as the full BLAKE2b hash requires (the ambiguity is caused by the fact that this conversion is specified as part of the full hash, not `F`).

There were two reasons why we decided having `m` and `h` accepted as `bytes32`:
1. Byte array does not have endianness so we did not want to enforce it on the BLAKE2b code using F precompile,
2. EVM uses 256-bit words, so `bytes32` is memory-efficient. 

What we did not take into account is what you pointed out in your comment - `h` and `m` inputs to `F` should be read as little-endian. As you have already seen in our implementation, we read them as big-endian just like EVM. I know see that point 1 does not make any sense.

I think the best option here is to leave `h` and `m` as we have them now: `bytes32[2] memory h, bytes32[4] memory m` but make it clear in the EIP that they are arrays of unsigned 32-bit words of little-endian bit order.
 
> For the `uint64` values, on the other hand, it's totally unclear what the intent is. Note that RFC 7693 specifies `t` as a single integer, and unpacks it into words within the definition of `F`. I would imagine that users of this API would try to just increment `t[0]` for each block, and then using `abi.encodePacked` here would be plain wrong (the byte order would be swapped).

Agreed. I think that's another reason (next to EVM memory model) why we should keep `h` and `m` as `bytes32`.

The RFC says `t` is a `2w-bit offset counter` and I assume that it's 2-word counter of little endian interpretation according to [2.4.  Little-Endian Interpretation of Words as Bytes](https://tools.ietf.org/html/rfc7693#section-2.4). To avoid confusion, maybe it makes sense to accept `t` as `bytes16` and make it clear in the EIP it's unsigned 16-byte word of little-endian bit order: `[ 8 bytes for t0 ][ 8 bytes for t1 ]`.

> As well as clarifying this in the text, I strongly suggest adding test vectors (which should include `t` values >= 264).

We had a problem with test vectors for `F` - the RFC and all other official resources give test vectors for `BLAKE2b` as a whole. The test vectors we used in the `F` implementation and for benchmarks were generated from input/output pairs of execution of `F` using official test vectors for `BLAKE2b` as a whole. I do not have any better idea of how to generate something valuable and not entirely detached from the official documentation. If this method sounds good to you, we can add those test vectors to EIP, making sure they are little-endian and include `t >= 264`.

Reply:
> I think the best option here is to leave h and m as we have them now: bytes32[2] memory h, bytes32[4] memory m but make it clear in the EIP that they are arrays of unsigned 32-bit words of little-endian bit order.

Byte order, not bit order. (There's no defined ordering of bits within bytes here.)

> To avoid confusion, maybe it makes sense to accept `t` as `bytes16` and make it clear in the EIP it's unsigned 16-byte word of little-endian bit order: `[ 8 bytes for t0 ][ 8 bytes for t1 ]`.

Again you mean byte order, but yes that makes sense. It's a little harder to increment a byte array representation but not that much harder (especially as many applications will be limited to less than 256 blocks).

https://tools.ietf.org/html/rfc7693#appendix-A gives internal test vectors for the F function of BLAKE2b-512. Each 32-bit word in the internal state is presented as a hexadecimal integer, i.e. with the most significant hexadecimal digit on the left.

Reply:
Note that if you keep the number of rounds encoded in big endian, that needs to be really clearly documented, otherwise people might assume it has the same byte order as the other fields.

Reply:
@pdyraga @mhluongo was there any progress made on this question?

Reply:
Sorry, @mhluongo is on vacation and I was busy with travelling. I'll update the EIP next week with suggestions from @daira, including test vectors.

Reply:
@daira I think this is how precompile test vector should look like based on BLAKE2b test vector from Appendix A (https://tools.ietf.org/html/rfc7693#appendix-A). Please let me know if it makes sense to you. 

```
INPUT:
0000000c48c9bdf267e6096a3ba7ca8485ae67bb2bf894fe72f36e3cf1361d5f3af54fa5d182e6ad7f520e511f6c3e2b8c68059b6bbd41fbabd9831f79217e1319cde05b61626300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000001
```

OUTPUT:
```
ba80a53f981c4d0d6a2797b69f12f6e94c212f14685ac4b74b12bb6fdbffa2d17d87c5392aab792dc252d5de4533cc9518d38aa8dbf1925ab92386edd4009923
```

Reply:
See 64840d1 for updated information about endianness.